app/server/lib/AttachmentFileManager.ts

```diff
@@ -0,0 +1,180 @@
+}
+
+export interface AddFileResult {
+  fileIdent: string;
+  isNewFile: boolean;
+}
+
+export class StoresNotConfiguredError extends Error {
+  constructor() {
+    super('Attempted to access a file store, but AttachmentFileManager was initialized without store access');
+  }
+}
+
+export class StoreNotAvailableError extends Error {
+  public readonly storeId: AttachmentStoreId;
+
+  constructor(storeId: AttachmentStoreId) {
+    super(`Store '${storeId}' is not a valid and available store`);
+    this.storeId = storeId;
+  }
+}
+
+
+interface AttachmentFileManagerLogInfo {
+  fileIdent?: string;
+  storeId?: string;
+}
+
+/**
+ * Instantiated on a per-document to basis to provide a document with access to its attachments.
+ * Handles attachment uploading / fetching, as well as trying to ensure consistency with the local document database,
+ * which tracks attachments and where they're stored.
+ *
+ * This class should prevent the document code from having to worry about accessing the underlying stores.
+ */
+export class AttachmentFileManager implements IAttachmentFileManager {
+  // _docPoolId is a critical point for security. Documents with a common pool id can access each others' attachments.
+  private readonly _docPoolId: DocPoolId | null;
+  private readonly _docName: string;
+  private _log = new LogMethods(
+    "AttachmentFileManager ",
+    (logInfo: AttachmentFileManagerLogInfo) => this._getLogMeta(logInfo)
+  );
+
+  /**
+   * @param _docStorage - Storage of this manager's document.
+   * @param _storeProvider - Allows instantiating of stores. Should be provided except in test scenarios.
+   * @param _docInfo - The document this manager is for. Should be provided except in test scenarios.
+   */
+  constructor(
+    private _docStorage: DocStorage,
+    private _storeProvider: IAttachmentStoreProvider | undefined,
+    _docInfo: AttachmentStoreDocInfo | undefined,
+  ) {
+    this._docName = _docStorage.docName;
+    this._docPoolId = _docInfo ? getDocPoolIdFromDocInfo(_docInfo) : null;
+  }
+
+  public async addFile(
+    storeId: AttachmentStoreId | undefined,
+    fileExtension: string,
+    fileData: Buffer
+  ): Promise<AddFileResult> {
+    const fileIdent = await this._getFileIdentifier(fileExtension, Readable.from(fileData));
+    this._log.info({ fileIdent, storeId }, `adding file to ${storeId ? "external" : "document"} storage`);
+    if (storeId === undefined) {
+      return this._addFileToLocalStorage(fileIdent, fileData);
+    }
+    const store = await this._getStore(storeId);
+    if (!store) {
+      this._log.info({ fileIdent, storeId }, "tried to fetch attachment from an unavailable store");
+      throw new StoreNotAvailableError(storeId);
+    }
+    return this._addFileToAttachmentStore(store, fileIdent, fileData);
+  }
+
+  public async getFileData(fileIdent: string): Promise<Buffer | null> {
+    this._log.debug({ fileIdent }, "retrieving file data");
```

Comment:
I'm concerned this logging might be a bit much. When a document is opened, it loads attachment previews. That results in 1 or 2 log lines for every attachment in a document.

Reply:
That might indeed be a bit much.

Reply:
Removed 1 of the log lines - the other one is quite useful for debugging as it mentions the attachment pool, and which storage (document / external) the file is going into.